### PR TITLE
[#477] normalWS exposed when tessellation and normal mapping is used

### PR DESCRIPTION
**Pull request by johang88**
# PR Details

Tessellation (with displacment) requires normalWS to be available, this is only the case if normal mapping is not used. This change exposes normalWS in the various normal mapping scenarios. 

The `GenerateNormal_VS` implementation from `NormalFromMesh` is used in all cases.

## Related Issue

#477 This solves one of the reported issues

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: sources/engine/Stride.Rendering/Rendering/Core/NormalFromNormalMappingTessellation.sdsl [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+// Distributed under the MIT license. See the LICENSE.md file in the project root for more information.
+/// <summary>
+/// Calculates the normal  the normals from a normal map.
+/// </summary>
+shader NormalFromNormalMappingTessellation : NormalFromNormalMapping
+{
+    override stage void GenerateNormal_VS()
+    {
+        // Perform normal generation at the end in case meshNormal is modified
+        streams.meshNormalWS = mul(streams.meshNormal, (float3x3)WorldInverseTranspose);    // TODO: PERFORMANCE: Normalization required?
```
Comment (Eideren) [inline sources/engine/Stride.Rendering/Rendering/Core/NormalFromNormalMappingTessellation.sdsl:11]: Is that todo still valid ?

File: sources/engine/Stride.Rendering/Rendering/Core/NormalFromNormalMappingTessellationInstanced.sdsl [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+// Distributed under the MIT license. See the LICENSE.md file in the project root for more information.
+/// <summary>
+/// Calculates the normal  the normals from a normal map.
+/// </summary>
+shader NormalFromNormalMappingTessellationInstanced : NormalFromNormalMappingInstanced
+{
+    override stage void GenerateNormal_VS()
+    {
+        // Perform normal generation at the end in case meshNormal is modified
+        streams.meshNormalWS = mul(streams.meshNormal, (float3x3)WorldInverseTranspose);    // TODO: PERFORMANCE: Normalization required?
```
Comment (Eideren) [inline sources/engine/Stride.Rendering/Rendering/Core/NormalFromNormalMappingTessellationInstanced.sdsl:11]: It's here as well

File: sources/engine/Stride.Rendering/Rendering/Skinning/NormalVSSkinningNormalMappingTessellation.sdsl [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+// Distributed under the MIT license. See the LICENSE.md file in the project root for more information.
+/// <summary>
+/// Calculates the normal  the normals from a normal map.
+/// </summary>
+shader NormalVSSkinningNormalMappingTessellation : NormalVSSkinningNormalMapping
+{
+    override stage void GenerateNormal_VS()
+    {
+        // Perform normal generation at the end in case meshNormal is modified
+        streams.meshNormalWS = mul(streams.meshNormal, (float3x3)WorldInverseTranspose);    // TODO: PERFORMANCE: Normalization required?
```
Comment (Eideren) [inline sources/engine/Stride.Rendering/Rendering/Skinning/NormalVSSkinningNormalMappingTessellation.sdsl:11]: yep

**Review by Eideren — COMMENTED** [review]
Looks pretty good, just one small thing to note

File: sources/engine/Stride.Rendering/Rendering/Core/NormalFromNormalMappingTessellation.sdsl (hunk 1 above)
Comment (johang88) [reply sources/engine/Stride.Rendering/Rendering/Core/NormalFromNormalMappingTessellation.sdsl:11]: Not sure if the todo is valid, its present in `NormalFromMesh` which is why I kept it.

**Comment by Eideren** [comment]
@johang88 Give me a heads up when you're ready to merge this in :)

**Comment by johang88** [comment]
WIll do, found some issues with instancing (it crashes) that I have to sort out first :)

**Comment by johang88** [comment]
There we go, instanced tessellation works without crashing now and looks correct. Had to change the `GetInstanceWorld` and `GetInstanceWorldInverse` methods as I was unable to find a way to get the shader compiler to generate distinct versions for each shader stage otherwise, which resulted in it generating a single function taking stream inputs from both VS and PS stages.

If this is possible to do somehow then we can change it back and get slightly cleaner shader code :)

**Comment by Eideren** [comment]
No clue, what do you think @tebjan ?

**Comment by tebjan** [comment]
@johang88 that looks correct. 

In general, I wonder why the effect file needs all this if/else constructs. my feeling is that this could be solved more "general" with permutation keys. @xen2 what do you think? my concern is, that for the next feature there would be another if/else, which kind of defies the purpose of dynamic composition...

Anyway, this doesn't influence the functionality, at the moment. So is there a test project somewhere that demonstrates the working features?

**Comment by johang88** [comment]
I agree, it' getting a bit unwieldy, especially in this case where it does not really add much. Some kind of inter-stage dependency analysis would work quite well in this case, ie always calculate `normalWS` in the vertex stage, the optimizer then removes it if not in use in subsequent shader stages.

**Comment by johang88** [comment]
Test project available here here https://github.com/johang88/Stride-Tessellation

**Comment by tebjan** [comment]
@johang88 I had some time to test the changes. it does indeed fix the cobblestone material used in your test project.

then I tried it with the instancing test project and added tesselation and displacement to a material but it didn't work there. as far as I remember, it wasn't working before your PR either, but I think I'll let you know, maybe the problem is related:
image

**Comment by johang88** [comment]
@tebjan I will check it out, not sure if I tested skinned instanced meshes so there might be an issue there. You are correct in that it did not work before but would be nice to fix all edge cases before merging.

**Comment by tebjan** [comment]
right, this could be related to skinning. if you find the time, it would be great to see a fix there, but this could also be another PR. so from my side, this PR is approved.

**Comment by johang88** [comment]
I did some test with the instancing project and it seems to work fine. So I think we are ready to merge.

image

**Comment by tebjan** [comment]
indeed, it works correctly here too with the latest beta build. great, thanks for the PR!

**Pull request merged**